### PR TITLE
fix: improved support for cjs and cts modules

### PR DESCRIPTION
**Pull request by dsherret**
* cts support
* better cjs/cts type checking
* deno compile cjs/cts support
* More efficient detect cjs (going towards stabilization)
  * Determination of whether .js, .ts, .jsx, or .tsx is cjs or esm is only done after loading
* https://github.com/denoland/deno_graph/pull/540
* https://github.com/denoland/deno_ast/pull/281
* https://github.com/denoland/deno_ast/pull/280
* https://github.com/denoland/deno_ast/pull/284

Closes https://github.com/denoland/deno/issues/26651
Closes https://github.com/denoland/deno/issues/25498

### CODE REVIEW

File: Cargo.toml
Comment (dsherret) [inline Cargo.toml:1]: I'm sorry, this PR is hard to split up into multiple smaller ones. There's so much stuff that's intertwined.

The main themes are using "program" more and determining if a `js`, `jsx`, `ts`, or `tsx` file is cjs or esm always only happens after loading it now.

File: ext/node/lib.rs
Comment (dsherret) [inline ext/node/lib.rs:1]: cc @arnauorriols on these changes and this PR in general. There's now a `NodeRequireLoader` trait for loading files with `require(...)` and `PackageJsonResolver` struct that searches for package.jsons. Let me know if you'd like a walkthrough of the changes.

File: cli/module_loader.rs [hunk 1]
```diff
@@ -607,20 +665,40 @@ impl<TGraphContainer: ModuleGraphContainer>
         source,
         media_type,
         specifier,
+        is_script,
         ..
       })) => {
+        // todo(dsherret): revert in https://github.com/denoland/deno/pull/26439
```
Comment (bartlomieju) [inline cli/module_loader.rs:671]: Make sure to link it back in the PR so it's not missed

File: cli/module_loader.rs [hunk 2]
```diff
@@ -900,3 +1028,79 @@ impl ModuleGraphUpdatePermit for WorkerModuleGraphUpdatePermit {
...
+      // allow reading if it's in the module graph
+      if self.graph_container.graph().get(&url).is_some() {
+        return Ok(std::borrow::Cow::Borrowed(path));
+      }
+    }
+    self.npm_resolver.ensure_read_permission(permissions, path)
+  }
+
+  fn load_text_file_lossy(&self, path: &Path) -> Result<String, AnyError> {
+    // todo(dsherret): use the preloaded module from the graph if available?
```
Comment (bartlomieju) [inline cli/module_loader.rs:1079]: Is this actionable? Maybe open an issue about it?

**Review by bartlomieju — COMMENTED** [review]
First pass, ended at `cli/resolver.rs` because the rest of files doesn't load in the browser :(

File: cli/tools/compile.rs [hunk 3]
```diff
@@ -53,16 +53,6 @@ pub async fn compile(
     );
   }
 
-  if cli_options.unstable_detect_cjs() {
```
Comment (bartlomieju) [inline cli/tools/compile.rs:56]: Is this on purpose? Shouldn't we land https://github.com/denoland/deno/pull/26439 first?

File: cli/tsc/mod.rs [hunk 4]
```diff
@@ -489,25 +559,22 @@ fn op_load_inner(
...
       .with_context(|| format!("Unable to load {}", file_path.display()))?;
-    Ok(code)
+    let code: Arc<str> = code.into();
+    *is_cjs = npm_state
+      .map(|npm_state| {
+        npm_state.cjs_tracker.is_cjs(specifier, *media_type, &code)
+      })
+      .unwrap_or(false);
+    // todo(dsherret): how to avoid cloning here?
+    Ok(code.to_string())
```
Comment (bartlomieju) [inline cli/tsc/mod.rs:577]: We can make `OpLoadResponse` accept `Arc<str>` and then the `op_load` would return a `v8::Array` that would manually serialize to `v8::String`

File: ext/node/lib.rs [hunk 5]
```diff
@@ -761,33 +761,16 @@ deno_core::extension!(deno_node,
...
-  main: bool,
-  inspect_brk: bool,
-) -> Result<(), AnyError> {
-  fn escape_for_single_quote_string(text: &str) -> String {
-    text.replace('\\', r"\\").replace('\'', r"\'")
-  }
-
-  let source_code = format!(
-    r#"(function loadCjsModule(moduleName, isMain, inspectBrk) {{
-      Deno[Deno.internal].node.loadCjsModule(moduleName, isMain, inspectBrk);
```
Comment (bartlomieju) [inline ext/node/lib.rs:776]: `loadCjsModule` can probably be removed now. I can do that in a follow up

File: cli/args/mod.rs [hunk 6]
```diff
@@ -201,6 +201,8 @@ pub fn ts_config_to_transpile_and_emit_options(
       precompile_jsx_dynamic_props: None,
       transform_jsx,
       var_decl_imports: false,
+      // todo(dsherret): support verbatim_module_syntax here properly
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/args/mod.rs:204]: No. Opened https://github.com/denoland/deno/issues/26678

File: cli/cache/module_info.rs [hunk 7]
```diff
@@ -202,7 +276,7 @@ fn serialize_media_type(media_type: MediaType) -> i64 {
     Tsx => 11,
     Json => 12,
     Wasm => 13,
-    TsBuildInfo => 14,
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/cache/module_info.rs:205]: We never use this media type. We cache bust with very deno version change.

File: cli/lsp/resolver.rs [hunk 8]
```diff
@@ -496,38 +562,67 @@ async fn create_npm_resolver(
...
+fn create_cjs_tracker(
+  in_npm_pkg_checker: Arc<dyn InNpmPackageChecker>,
+  pkg_json_resolver: Arc<PackageJsonResolver>,
+) -> Arc<CjsTracker> {
+  Arc::new(CjsTracker::new(
+    in_npm_pkg_checker,
+    pkg_json_resolver,
+    CjsTrackerOptions {
+      // todo(dsherret): support in the lsp by stabilizing the feature
+      // so that we don't have to pipe the config in here
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/lsp/resolver.rs:582]: The tracking issue has it: https://github.com/denoland/deno/issues/26225 (just updated it to be more clear)

File: cli/main.rs [hunk 9]
```diff
@@ -135,7 +135,7 @@ async fn run_subcommand(flags: Arc<Flags>) -> Result<i32, AnyError> {
       tools::compile::compile(flags, compile_flags).await
     }),
     DenoSubcommand::Coverage(coverage_flags) => spawn_subcommand(async {
-      tools::coverage::cover_files(flags, coverage_flags).await
+      tools::coverage::cover_files(flags, coverage_flags)
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/main.rs:138]: Yes. It's now only creates an InNpmPackageChecker which doesn't require async initialization.

File: cli/module_loader.rs [hunk 10]
```diff
@@ -578,6 +628,14 @@ impl<TGraphContainer: ModuleGraphContainer>
           media_type,
         }))
       }
+      Some(CodeOrDeferredEmit::Cjs { .. }) => {
+        self.parsed_source_cache.free(specifier);
+
+        // todo(dsherret): to make this work, we should probably just
+        // rely on the CJS export cache. At the moment this is hard because
+        // cjs export analysis is only async
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/module_loader.rs:636]: https://github.com/denoland/deno/issues/26679

File: cli/module_loader.rs (hunk 2 above)
Comment (dsherret) [reply cli/module_loader.rs:1079]: It's maybe a performance optimization, but it also might introduce undesirable behaviour. I'm not sure it's worth tracking in an issue.

File: cli/tools/compile.rs (hunk 3 above)
Comment (dsherret) [reply cli/tools/compile.rs:56]: Yes. It works in deno compile now so I got rid of the warning saying it doesn't work.

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**